dev/com.ibm.ws.kernel.feature.core/src/com/ibm/ws/kernel/feature/internal/FeatureIntrospector.java

```diff
@@ -0,0 +1,94 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.kernel.feature.internal;
+
+import java.io.PrintWriter;
+import java.util.Collection;
+import java.util.Set;
+
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+import com.ibm.ws.kernel.feature.FeatureDefinition;
+import com.ibm.ws.kernel.feature.FeatureProvisioner;
+import com.ibm.ws.kernel.feature.provisioning.FeatureResource;
+import com.ibm.ws.kernel.feature.provisioning.ProvisioningFeatureDefinition;
+import com.ibm.wsspi.logging.Introspector;
+
+/**
+ * Introspector implementation that prints server feature metadata
+ * in the event of a dump
+ */
+
+@Component
+public class FeatureIntrospector implements Introspector {
+
+    private final FeatureProvisioner provisioner;
+
+    @Activate
+    public FeatureIntrospector(@Reference FeatureProvisioner provisioner) {
+        this.provisioner = provisioner;
+    }
+
+    @Override
+    public String getIntrospectorName() {
+        return "FeatureIntrospector";
+    }
+
+    @Override
+    public String getIntrospectorDescription() {
+        return "Lists all the features enabled in the server.";
+    }
+
+    /**
+     * Requests the set of features that are installed on the server and outputs
+     * relevant data to the log file
+     */
+    @Override
+    public void introspect(PrintWriter out) throws Exception {
+        Set<String> features = this.provisioner.getInstalledFeatures();
+        for (String feature : features) {
+            FeatureDefinition featureDef = this.provisioner.getFeatureDefinition(feature);
+            out.println(featureDef.getFeatureName() + "{");
+            out.println("    Visibility: " + featureDef.getVisibility());
+            out.println("    Version: " + featureDef.getVersion());
+            out.println("    Simple Name: " + featureDef.getFeatureName());
+            out.println("    Symbolic Name: " + featureDef.getSymbolicName());
+            if (featureDef instanceof ProvisioningFeatureDefinition) {
+                ProvisioningFeatureDefinition proFeatDef = (ProvisioningFeatureDefinition) featureDef;
+                String bundleRepType = proFeatDef.getBundleRepositoryType();
+                out.println("    Bundle Repository Type: " + (bundleRepType == null || bundleRepType.isEmpty() ? "core" : bundleRepType));
+                out.println("    Auto Feature: " + proFeatDef.isAutoFeature());
+                out.println("    Singleton: " + proFeatDef.isSingleton());
+                out.print("    Superseded By: ");
+                try {
+                    out.println(proFeatDef.getSupersededBy()); //Throws Exception if called outside of provisioning operation
```

Comment:
I think we can work around this by using internals of this project/bundle.  First you will need to move this class into the package `com.ibm.ws.kernel.feature.internal.subsystem` so that you can use package private methods of that package.  Then you can do something like how `com.ibm.ws.kernel.feature.internal.subsystem.FeatureRepository.LibertyFeatureServiceFactory.getService(Bundle, ServiceRegistration<LibertyFeature>)` does to set the details so we can get this information.